packages/vm/src/imports.rs

```diff
@@ -235,6 +258,34 @@ fn do_humanize_address<A: Api, S: Storage, Q: Querier>(
     }
 }
 
+fn do_verify_secp256k1<A: Api, S: Storage, Q: Querier>(
+    env: &Environment<A, S, Q>,
+    hash_ptr: u32,
+    signature_ptr: u32,
+    pubkey_ptr: u32,
+) -> VmResult<u32> {
+    let hash = Binary(read_region(&env.memory(), hash_ptr, LENGTH_SHA256_HASH)?);
+
+    let signature = Binary(read_region(
+        &env.memory(),
+        signature_ptr,
+        MAX_LENGTH_SIGNATURE,
+    )?);
+
+    let pubkey = Binary(read_region(&env.memory(), pubkey_ptr, MAX_LENGTH_PUBKEY)?);
+
+    let result = secp256k1_verify(&hash, &signature, &pubkey);
+    let gas_info = GasInfo::with_cost(GAS_COST_VERIFY_SECP256K1_SIGNATURE);
+    process_gas_info::<A, S, Q>(env, gas_info)?;
+    match result {
+        Ok(()) => Ok(0),
+        Err(VmError::CryptoErr { msg, .. }) => {
+            Ok(write_to_contract::<A, S, Q>(env, msg.as_bytes())?)
```

Comment:
I'd really like to avoid returning error messages to the contract. In practice they are not processed anyways. I think we have two cathegories of errors:

1. Verification failure
2. Something went horribly wrong

For 1. we can return a boolean result to the contract. For 2. we can let `do_verify_secp256k1` error, which terminates the execution and prints an error message to the transaction log which can be inspected for debugging.

Reply:
OK. Please take a look at the latest version.